### ISSUE TITLE
No compiler error when returning object with extra properties that are unspecified in interface (sometimes)

### ISSUE THREAD

**Issue by ianhenderson** [Microsoft/TypeScript#15212]
**TypeScript Version:**  2.2.1

**Code**

```ts
interface MyType {
    value: string
}

function MyTypeMaker(value: string): MyType {
    return {
        value,
        foo: 'bar' // <-- As expected, this line shows an error in IDE...
    }
}

function MyTypeMaker2(value: string): MyType {
    const out: MyType = {
        value,
        foo: 'bar' // <-- As expected, this line shows an error in IDE as well...
    }
    return out
}

function MyTypeMaker3(value: string): MyType {
    const out = {
        value,
        foo: 'bar' // <-- But this is allowed...
    }
    return out
}

```

**Expected behavior:**

Since no optional properties are specified on the `MyType` interface I would have expected tsc to also detect an error in `MyTypeMaker3`, as in the other two examples.  Am I missing something..?

**Actual behavior:**
No problem is indicated.

Apologies if this is addressed in another issue -- I tried searching but couldn't find one.

Thank you in advance!

**Comment by j-oliveras** [comment]
[This work as expected](https://www.typescriptlang.org/docs/handbook/type-compatibility.html), in third function, `out` has a inferred type `{ value: string, foo: string }` that is compatible with `MyType`.

**Comment by RyanCavanaugh** [comment]
This is basically the reverse of http://stackoverflow.com/questions/22077023/why-cant-i-indirectly-return-an-object-literal-to-satisfy-an-index-signature-re

**Comment by ianhenderson** [comment]
Thanks for the clarification @j-oliveras and @RyanCavanaugh 👍
